collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManager.java

```diff
@@ -0,0 +1,296 @@
+ *
+ * For location audit logging, manages all the audit logging as well as fetching the location using
+ * Google Play Services.
+ *
+ * {@link org.odk.collect.android.logic.actions.setlocation.CollectSetLocationAction} fetches
+ * location for odk:setlocation actions.
+ *
+ * The implementation uses a state machine concept. Public method represent user or system actions
+ * that clients of this class capture. Based on those actions and various preconditions (Google Play
+ * Services available, location permissions granted, etc), the manager's state changes.
+ */
+public class BackgroundLocationManager implements LocationClient.LocationClientListener, LocationListener {
+    @NonNull
+    private BackgroundLocationState currentState;
+
+    @NonNull
+    private LocationClient locationClient;
+
+    @NonNull
+    private LocationListener locationListener;
+
+    @NonNull
+    private BackgroundLocationHelper helper;
+
+    BackgroundLocationManager(LocationClient locationClient, BackgroundLocationHelper helper) {
+        currentState = BackgroundLocationState.NO_BACKGROUND_LOCATION_NEEDED;
+        this.locationClient = locationClient;
+
+        this.helper = helper;
+
+        this.locationListener = this;
+    }
+
+    void formFinishedLoading() {
+        switch (currentState) {
+            case NO_BACKGROUND_LOCATION_NEEDED:
+                if (helper.currentFormCollectsBackgroundLocation()) {
+                    currentState = BackgroundLocationState.PENDING_PRECONDITION_CHECKS;
+                }
+        }
+    }
+
+    BackgroundLocationMessage activityDisplayed() {
+        switch (currentState) {
+            case PENDING_PRECONDITION_CHECKS:
+                // Separate out user message so that any failed precondition is written to the audit
+                // log. If Play Services are unavailable AND the location preference is disabled,
+                // show the Play Services unavailable message only.
+                BackgroundLocationMessage userMessage = null;
+
+                if (!helper.isBackgroundLocationPreferenceEnabled()) {
+                    helper.logAuditEvent(AuditEvent.AuditEventType.LOCATION_TRACKING_DISABLED);
+                    userMessage = BackgroundLocationMessage.LOCATION_PREF_DISABLED;
+                }
+
+                if (!helper.arePlayServicesAvailable()) {
+                    helper.logAuditEvent(AuditEvent.AuditEventType.GOOGLE_PLAY_SERVICES_NOT_AVAILABLE);
+                    userMessage = BackgroundLocationMessage.PLAY_SERVICES_UNAVAILABLE;
+                }
+
+                if (userMessage == null) {
+                    helper.logAuditEvent(AuditEvent.AuditEventType.LOCATION_TRACKING_ENABLED);
+                    currentState = BackgroundLocationState.PENDING_PERMISSION_CHECK;
+                } else {
+                    currentState = BackgroundLocationState.STOPPED;
+                }
+
+                return userMessage;
+
+            case RECEIVING_LOCATIONS:
```

Comment:
Running `BackgroundLocationManagerTest` with coverage it doesn't look like this case gets hit by any of the tests

Reply:
Oh dear, that's because it was totally wrong and didn't represent a real case anymore. 😫 8b93f22 should address that.